connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ParseStructByRegex.java

```diff
@@ -0,0 +1,224 @@
+import java.util.regex.Pattern;
+
+public abstract class ParseStructByRegex<R extends ConnectRecord<R>> implements Transformation<R> {
+    public static final String OVERVIEW_DOC = "Generate key/value Struct objects supported by ordered Regex Group"
+        + "<p/>Use the concrete transformation type designed for the record key (<code>" + Key.class.getName() + "</code>) "
+        + "or value (<code>" + Value.class.getName() + "</code>).";
+
+    private interface ConfigName {
+        String REGEX = "regex";
+        String MAPPING_KEY = "mapping";
+    }
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef()
+        .define(ConfigName.REGEX, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE, new GroupRegexValidator(), ConfigDef.Importance.MEDIUM,
+            "String Regex Group Pattern.")
+        .define(ConfigName.MAPPING_KEY, ConfigDef.Type.LIST, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Importance.MEDIUM,
+            "Ordered Regex Group Mapping Keys");
+
+    private static final String PURPOSE = "Parse Struct by regex group mapping";
+
+    private String pattern;
+    private List<String> fieldKeys;
+
+    private Cache<Schema, Schema> schemaUpdateCache;
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        final AbstractConfig config = new AbstractConfig(CONFIG_DEF, configs, false);
+        pattern = config.getString(ConfigName.REGEX);
+        fieldKeys = config.getList(ConfigName.MAPPING_KEY);
+
+        if (fieldKeys.isEmpty()) {
+            throw new ConfigException("pattern must not be empty");
+        }
+        schemaUpdateCache = new SynchronizedCache<>(new LRUCache<>(16));
+    }
+
+    @Override
+    public R apply(R record) {
+        if (operatingSchema(record) == null) {
+            return applySchemaless(record);
+        } else {
+            return applyWithSchema(record);
+        }
+    }
+
+    private R applySchemaless(R record) {
+        Map<String, Object> resultMap;
+
+        String inputTargetStr;
+        if (operatingValue(record) instanceof String) {
+            inputTargetStr = (String) operatingValue(record);
+            resultMap = parseRegexGroupToStringMap(inputTargetStr);
+        } else {
+            throw new DataException("only plain string value support with ParseStructByRegex");
+        }
+
+        return newRecord(record, null, resultMap);
+    }
+
+    private R applyWithSchema(R record) {
+        Schema updatedSchema;
+        Struct updatedStruct;
+
+        String inputTargetStr;
+        if (operatingValue(record) instanceof String) {
+            inputTargetStr = (String) operatingValue(record);
+            Map<String, Object> newEntryMap = parseRegexGroupToStringMap(inputTargetStr);
+            updatedSchema = newSchema(newEntryMap);
+            updatedStruct = newStruct(newEntryMap, updatedSchema);
+        } else {
+            throw new DataException("only plain string value support with ParseStructByRegex");
```

Comment:
In order to avoid repeating the same validation logic in `applySchemaless` and `applyWithSchema`, we could extract it into a private method.